massa-protocol-worker/src/connectivity.rs

```diff
@@ -310,35 +314,58 @@ pub(crate) fn start_connectivity_thread(
-                                                if name == category && category_infos > &mut 0 {
-                                                        addresses_to_connect.push(*addr);
-                                                        *category_infos -= 1;
-                                                }
-                                            }
-                                        } else if slot_default_category > 0 &&  !addresses_to_connect.contains(addr) {
-                                            addresses_to_connect.push(*addr);
-                                            slot_default_category -= 1;
-                                        }
+                                        addresses_can_connect.push((*addr, connection_metadata, category_found));
+                                    }
+                                }
+                            }
+                        }
 
-                                            // IF all slots are filled, stop
-                                        if slot_default_category == 0 && slots_per_category.iter().all(|(_, slots)| *slots == 0) {
-                                            break;
+                        // Sort addresses using the metadata
+                        addresses_can_connect.sort_by(|a, b| a.1.cmp(&b.1));
+
+                        let mut addresses_to_connect = vec![];
+                        for (addr, _, category) in addresses_can_connect.iter() {
+                            match category {
+                                Some(cat) => {
+                                    for (name, category_infos) in &mut slots_per_category {
+                                        if name == *cat && category_infos > &mut 0 {
+                                            addresses_to_connect.push(*addr);
+                                            *category_infos -= 1;
                                         }
                                     }
                                 }
+                                None if !addresses_to_connect.contains(&addr) => {
+                                    slot_default_category -= 1;
+                                    addresses_to_connect.push(*addr);
+                                }
+                                _ => continue,
+                            }
+
+                            // IF all slots are filled, stop
+                            if slot_default_category == 0 && slots_per_category.iter().all(|(_, slots)| *slots == 0) {
+                                break;
                             }
                         }
 
                         for addr in addresses_to_connect {
                             info!("Trying to connect to addr {}", addr);
 
-                            peer_db.write().try_connect_history.insert(addr, MassaTime::now().unwrap());
+                            let conn_res = network_controller.try_connect(addr, config.timeout_connection.to_duration());
```

Comment:
"try_connect()" spawns a thread in PeerNet, so we don't have the real result here

Reply:
yeah :/ need to check but i think we have Ok if the thread was correctly spawned and an error in the other case

Reply:
Ok one thing I can do, is only do a `md.failure()` if the `try_connect` thread fails to start, and in the `MassaHandshake::perform_handshake` function, I call `md.success()` or `md.failure()` depending on how the handshake is going

Reply:
yes, for now i think it is the ~~only~~ right way